src/loaders/lottie/tvgLottieBuilder.cpp

```diff
@@ -1067,33 +1067,43 @@ void LottieBuilder::updateText(LottieLayer* layer, float frameNo)
 
                     //text range process
                     for (auto s = text->ranges.begin(); s < text->ranges.end(); ++s) {
-                        float start, end;
-                        (*s)->range(frameNo, float(totalChars), start, end);
-
                         auto basedIdx = idx;
                         if ((*s)->based == LottieTextRange::Based::CharsExcludingSpaces) basedIdx = idx - space;
                         else if ((*s)->based == LottieTextRange::Based::Words) basedIdx = line + space;
                         else if ((*s)->based == LottieTextRange::Based::Lines) basedIdx = line;
 
-                        if (basedIdx < start || basedIdx >= end) continue;
+                        auto f = (*s)->factor(frameNo, float(totalChars), (float)basedIdx);
+                        if (tvg::zero(f)) continue;
 
-                        translation = translation + (*s)->style.position(frameNo);
-                        scaling = scaling * (*s)->style.scale(frameNo) * 0.01f;
-                        rotation += (*s)->style.rotation(frameNo);
+                        translation = translation + f * (*s)->style.position(frameNo);
+                        scaling = scaling * (f * ((*s)->style.scale(frameNo) * 0.01f - Point{1.0f,1.0f}) + Point{1.0f,1.0f});
+                        rotation += f * (*s)->style.rotation(frameNo);
 
                         shape->opacity((*s)->style.opacity(frameNo));
 
                         auto color = (*s)->style.fillColor(frameNo);
-                        shape->fill(color.rgb[0], color.rgb[1], color.rgb[2], (*s)->style.fillOpacity(frameNo));
+                        if (f == 1.0f) shape->fill(color.rgb[0], color.rgb[1], color.rgb[2], (*s)->style.fillOpacity(frameNo));
+                        else {
+                            auto r = lerp<uint8_t>(doc.color.rgb[0], color.rgb[0], f);
+                            auto g = lerp<uint8_t>(doc.color.rgb[1], color.rgb[1], f);
+                            auto b = lerp<uint8_t>(doc.color.rgb[2], color.rgb[2], f);
+                            shape->fill(r, g, b, (*s)->style.fillOpacity(frameNo));
+                        }
```

Comment:
Isn't this, factor for fill color, this should multiply color by `f`? This looks like factor replaces RGB data instead.

Reply:
I don't understand what do you mean. 
so I double checked the blended colors, look fine

result AE vs this pr:
<img width="400" alt="Zrzut ekranu 2024-11-3 o 10 31 30" src="https://github.com/user-attachments/assets/21953ec9-4105-471c-abb0-da4187e6e3a0">

sample:
[smoothnessABC.json](https://github.com/user-attachments/files/17609131/smoothnessABC.json)

Reply:
@hermet please check now - color is now interpolated taking into account the pervious color, if was set

before last changes:
<img width="410" alt="Zrzut ekranu 2024-11-4 o 11 22 42" src="https://github.com/user-attachments/assets/713995ae-f609-4b14-86d6-cf5495f493a0">

after last change:
<img width="418" alt="Zrzut ekranu 2024-11-4 o 12 16 27" src="https://github.com/user-attachments/assets/8fe24eda-00cd-45b3-8082-fa9b925c8e99">

sample:
[smoothness2Ranges.json](https://github.com/user-attachments/files/17614017/smoothness2Ranges.json)